File: queue_01.py
'''
@Author: youngjam
@Date: 2020-04-23 14:12:56
@LastEditTime: 2020-04-23 14:43:20
@Description: 队列的数组实现《图解数据结构使用python实现》
@logs: 
@Environment: Ubuntu 18.04.4 LTS && gcc 7.5.0
'''
import sys

MAX = 10    # 声明队列大小

queue = [0] * MAX   # 声明队列
front = rear = -1   # 定义队列游标，rear指向队尾元素，front指向对首元素前一位置

choice = ''     # 程序控制字

while rear < MAX - 1 and choice != 'e':
    choice = input('[j] => join queue;  [l] => leave queue;  [e] => exit the program :\t')

    if choice == 'j':   # 输入元素加入队列
        val = int(input('input value:'))
        rear += 1
        queue[rear] = val
    
    elif choice == 'l':     # 队首元素出队

        if rear > front:    # 队列尚有元素
            front += 1
            print('the value of left: [%d]' % queue[front])
        
        else:   # 队列为空
            print('queue is empty!')
            sys.exit(0)


print('------------------------')
print('output the all value of queue:')

if rear == MAX - 1:
    print('queue is full!')

elif front >= rear:
    print('queue is empty!')

else:

     while rear > front:
         front += 1
         print('the value of left: [%d]' % queue[front])

print('------------------------')

File: c/queue_01.c
/*
 * @Author: youngjam
 * @Date: 2020-04-23 15:51:28
 * @LastEditTime: 2020-04-23 17:03:35
 * @Description: 基于数组的队列实现
 * @logs: 
 * @Environment: Ubuntu 18.04.4 LTS && gcc 7.5.0
 */
#include<stdio.h>
#include<malloc.h>
#include<stdlib.h>
#include<time.h>

#define MAXSIZE 20

typedef struct Queue {
    int data[MAXSIZE];
    int front;
    int rear;
}IntQueue;

/* 初始化队列的首尾游标 */
void InitQueue(IntQueue *p_queue)
{
    p_queue->front  = -1;
    p_queue->rear   = -1;
}   // end InitQueue

/* 将data从队尾加入队列 */
void Enqueue(IntQueue *p_queue, int data)
{
    if (p_queue->rear >= MAXSIZE - 1) {
        printf("queue is full!\n");
    }   // end if

    else {
        p_queue->rear++;
        p_queue->data[p_queue->rear] = data;
    }   // end else
}   // end Enqueue

/* 对首元素离开队列，并返回函数调用处 */
int Dequeue(IntQueue *p_queue)
{
    if (p_queue->front >= p_queue->rear) {
        printf("queue is empty!\n");
        return -1;
    }   // end if

    else {
        p_queue->front++;
        return p_queue->data[p_queue->front];
    }   // end else
}   // end Dequeue

int main(void)
{
    srand((int)time(NULL));

    IntQueue i_queue;       // 声明存放 int 类型队列

    InitQueue(&i_queue);

    char flag = 'e';

    while (flag != 'q') {

        int data = 0;

        printf("[e] => enqueue();  [d] => dequeue();  [q] => quit!\t");
        // flag = getchar();    /*==> error: 输出语句的'\n'会留在缓冲区赋给flag <==*/
        while ((flag = getchar()) == '\n');     // 消耗掉缓冲区 '\n' 
       
        switch (flag) {

            case 'e':
                data = rand() % 20;
                printf("value = %d\n", data);
                Enqueue(&i_queue, data);
                break;

            case 'd':
                data = Dequeue(&i_queue);
                printf("the value of left is: %d\n", data);
                break;

            default:
                break;
        }   // end switch
    }   // end while

    while (i_queue.rear > i_queue.front) {
        printf("the value of dequeue:\t%d\n", Dequeue(&i_queue));
    }   // end while
    
    return 0;
}   // end main()
